src/models/kinflated_gamma_poisson_model.py

```diff
@@ -0,0 +1,633 @@
+            dist = KInflatedGammaPoissonDistribution(alpha, beta, a)
+            score = self._chi2_distance(h, I, N, dist)
+            Imax = N * dist.expected_value()
+            if Imax < Imin:
+                score += (Imin - Imax) ** 2
+            return score
+
+        result = scipy.optimize.minimize(objective, p0, method="Nelder-Mead")
+
+        alpha, beta, N = result.x[:3]
+        a = result.x[3:]
+        dist = KInflatedGammaPoissonDistribution(alpha, beta, a)
+        score = self._chi2_distance(h, I, N, dist)
+
+        return N, dist, score
+
+    def print_fit_header(self):
+        if not self._debug:
+            return
+        print(
+            "{:15s} {:>7s} {:>7s} {:>8s} {:>8s} {:>2s}  distribution".format(
+                " ", "score", "N", "alpha", "beta", "n"
+            )
+        )
+
+    def print_fit(
+        self,
+        msg: str,
+        score: float,
+        N: float,
+        alpha: float,
+        beta: float,
+        a: List[float],
+    ):
+        if not self._debug:
+            return
+        dist = KInflatedGammaPoissonDistribution(alpha, beta, a)
+        dist_str = " ".join([f"{dist.pmf(i):7.4f}" for i in range(1, 10)])
+        print(
+            f"{msg:15s} {score:7.2f} {N:7.0f} {alpha:8.3f} {beta:8.3f} {len(a):2d} {dist_str}"
+        )
+
+    def _fit_point(
+        self, point: ReachPoint, Imin: float
+    ) -> Tuple[float, KInflatedGammaPoissonDistribution]:
+        """Chi-squared fit to histogram h.
+
+        Computes parameters alpha, beta, Imax, N such that the histogram hbar of
+        of expected frequencies minimizes the metric
+
+          chi2(h, hbar) = \sum_i (h[i] - hbar[i])^2 / hbar[i]
+
+        subject to the constraint the number of impressions in the fitted model
+        is at least Imin.
+
+        Args:
+          point:  The point to which the data is to be fit.
+          Imin: The minimum number of impressions that the resulting
+            fit should have.
+        Returns:
+          N, dist of the fit that was found, where N is the estimated audience size
+          and dist is the best fitting k-inflated Gamma Poisson distribution.
+          The expected number of impressions of this pair will be at least
+          Imin.
+        """
+        N0, dist = self._fit_exponential_poisson_model(point, Imin)
+        h = point.frequencies_with_kplus_bucket
+        I = point.impressions
+        score = self._chi2_distance(h, I, N0, dist)
+        self.print_fit_header()
+        self.print_fit("exp-poisson", score, N0, dist._alpha, dist._beta, dist._a)
+
+        if point.max_frequency == 1:
+            return N0, dist
+
+        N, dist, score = self._fit_histogram_fixed_length_a(
+            h, I, Imin, N0, dist._alpha, dist._beta, []
+        )
+        self.print_fit("start", score, N, dist._alpha, dist._beta, dist._a)
+
+        while scipy.stats.chi2.cdf(score, len(h)) > 0.95 and len(dist._a) < self._kmax:
```

Comment:
Is `scipy.stats.chi2.cdf(score, len(h))` always decreasing as len(a) increases?

Reply:
In this specific algorithm, the answer to that question is yes.  The reason is that each new iteration is initialized from the best values that were found in the previous iteration.  In other words, when a is extended in line 554, the new value that is added to a is the value of the Gamma-Poisson PMF at that point.  The optimizer is guaranteed to return a point where the objective function is no greater than the objective function at the starting point.  Consequently, the values of `score` are guaranteed to be decreasing.  Since `scipy.stats.chi2.cdf` is a monotonic function, it follows that `scipy.stats.chi2.cdf(score, len(h))` is guaranteed to decrease with each iteration.

Reply:
Got it!